### PR TITLE
NEW: Add Android backButtonLeavesApp to New Input

### PR DESCRIPTION
**Pull request by manugildev**
### Description
Add [`backButtonLeavesApp`](https://docs.unity3d.com/ScriptReference/Input-backButtonLeavesApp.html) support for New Input System. 

### Changes made
This PR creates a new `AndroidDevice` that is registered during initialization. This `AndroidDevice` represents a handheld device, and can be use to store device specific input functions that are not part of other input devices such as the keyboard, gamepad or sensors. 

We also added `SetCustomCommand` and `GetCustomCommand` used to be able to set and get respectively a payload between the InputSystem package and runtime. This is then used to set and get `backButtonLeavesApp` to the native runtime. This two new commands are general purpose commands, and can be used for any other input device.

`backButtonLeavesApp`, is just a boolean flag needed for [Predictive Back Gesture (PBG)](https://docs.unity3d.com/6000.0/Documentation/ScriptReference/PlayerSettings.Android-predictiveBackSupport.html) support to work on Android. This allows users to preview the destination or other result of a back gesture before fully completing it, allowing them to decide whether to continue or stay in the current view.

### Notes
This new API can be used as such:
```csharp
InputSystem.settings.android.backButtonLeavesApp = true;
```

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [x] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [x] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [x] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1914) <br/>All committers have signed the CLA.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/GetCustomCommand.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/GetCustomCommand.cs]: Its not clear to me what this field represents, would you mind documenting this field?

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/GetCustomCommand.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/GetCustomCommand.cs]: It is not clear to me what to expect from this field, also is "value" the most representative name for this field?

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs:43]: For how far back is this command supported in native run-time? I assume the command would fail with an error if not supported in backend and this defaults to false, correct?

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidSupport.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidSupport.cs]: Would suggest replacing this "AndroidDevice" string with `nameof(AndroidDevice)`

**Review by ekcoh — CHANGES_REQUESTED** [review]
@manugildev Thanks for submitting this PR. Would you mind filling in the PR template to aid the review process for this new functionality and API surface being proposed? Instructions for filling in the PR template are available in the PR itself but also here: https://github.com/Unity-Technologies/InputSystem/blob/develop/.github/pull_request_template.md
Clearly describing the purpose of the PR and motivating the changes made makes it simpler to understand and review.

In addition som initial questions:
- This is adding a new type of device, AndroidDevice that communicates with the native backend/run-time via IOCTL command GCC. However I do not see any Unity version dependent #ifdef checks present. Is this IOCTL command available back to Unity 2019 which the Input System package is still compliant with? If not, please guard the availability of this feature under appropriate compile-time version checks if intended to just return default if runtime is not implementing the IOCTL command and document from which Unity version this is expected to work, or guard the API under a certain UNITY version to control availability.
- Noted that CHANGELOG.md do not contain an addition under ADD section describing the addition, would it be possible to get that added?

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/GetCustomCommand.cs
Comment (todi1856) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/GetCustomCommand.cs]: GetCustomCommand/SetCustomCommnad are general purpose classes, the purposes of these classes - anyone can use them, if code/value is enough represent the flow of information.

Maybe this should be in generic place?

**Comment by todi1856** [comment]
P.S you probably need to some tests to validate SetCustomCommand/GetCustomCommand works fine

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs
Comment (todi1856) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs:36]: we should probably exit here without ApplySettings call if value matches m_BackButtonLeavesApp ?

```
if (value == m_BackButtonLeavesApp)
    return;
```

otherwise if you use it with stuff like IMGUI, you could end up calling ApplySettings every frame

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/GetCustomCommand.cs
Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/GetCustomCommand.cs]: renamed to payload https://github.com/Unity-Technologies/InputSystem/pull/1914/commits/30c57bc4277ce7695d89394cbb616824065f1d5c

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs
Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs:43]: added #ifdef guards: 31bedc2801ac313a4485864880a9b324ed2df6e6

sending IOCTL commands to runtime has no effect if there are no callbacks that picked them up and act on them.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs
Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs:36]: fee01d001061c0aab49e7bd8299f7103bd41b3b9

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs
Comment (todi1856) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs:41]: Does this getter ever called? I would assume you would need to use to initialize InputSystem.settings.android.backButtonLeavesApp, but not sure, can you comment on this?

Comment (todi1856) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs:41]: Maybe add a simple test in Assets/Tests/InputSystem/Plugins/AndroidTests.cs for InputSystem.settings.android.backButtonLeavesApp ? Not sure

Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs:40]: The getter is not being called, I will remove it.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs
Comment (todi1856) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs:18]: What's this? I see it's a public API, how would user use it ?

Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs:18]: Not sure how to get rid of this ButtonControl. InputDevices need to have AT LEAST one control. If not they throw an exception. In our case, this controls is not needed

**Review by todi1856 — APPROVED** [review]

**Comment by ritamerkl** [comment]
Add/change deprecation message to Input.bindings backButtonLeavesApp once ready to land.

File: Assets/Tests/InputSystem/CoreTests_Devices.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/CoreTests_Devices.cs]: Would prefer to not have unsafe on the whole test method but rather on individual unsafe operations when used in a test like this.

File: Assets/Tests/InputSystem/CoreTests_Devices.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/CoreTests_Devices.cs]: Would prefer to not have unsafe on the whole test method but rather on individual unsafe operations when used in a test like this.

File: Assets/Tests/InputSystem/Plugins/AndroidTests.cs
Comment (ekcoh) [inline Assets/Tests/InputSystem/Plugins/AndroidTests.cs]: Would prefer to not have unsafe on the whole test method but rather on individual unsafe operations when used in a test like this.

File: Packages/com.unity.inputsystem/InputSystem/Devices/Commands/GetCustomCommand.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Devices/Commands/GetCustomCommand.cs:15]: I do not see this FourCC code in InputFourCC.h in native code or did I overlooked something?

Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Devices/Commands/GetCustomCommand.cs:15]: (I recommend adding it there to avoid future collision)

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs:18]: @manugildev Could you provide some additional context regarding what exception/trace you did get?

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs:12]: Why would this be something Android specific when we advertise it as cross-platform on 3 operating systems for the UnityEngine.Input counterpart? https://docs.unity3d.com/ScriptReference/Input-backButtonLeavesApp.html

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs:29]: In UnityEngine.Input there is guidance about manually responding to back button, is something similar needed here maybe?

**Review by ekcoh — COMMENTED** [review]
Left another round of review comments. Generally I struggle a bit to see how this fits API-wise. It seems like we are transforming a cross-platform interface into a OS-specific here and introducing what I assume is supposed to be a non-removable device representing the system. 

Thanks for updating migration doc but I also think there is some guidance needed to explain how this boolean flag is expected to be used for the preview render feature. At the moment I do not fully understand how I should reason about this setting or what I need to do differently depending on how this flag is set. Happy to follow-up via DM or have a chat and maybe a demo that helps clarify this.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs
Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs:18]:
```
Could not create a device for 'AndroidDevice (Android)' (exception: System.InvalidOperationException: Control '/AndroidDevice' with layout 'AndroidDevice' has no size set and has no children to compute size from
 at UnityEngine.InputSystem.Layouts.InputDeviceBuilder.ComputeStateLayout (UnityEngine.InputSystem.InputControl control)
 at UnityEngine.InputSystem.Layouts.InputDeviceBuilder.InstantiateLayout (UnityEngine.InputSystem.Layouts.InputControlLayout layout, UnityEngine.InputSystem.Utilities.InternedString variants, UnityEngine.InputSystem.Utilities.InternedString name, UnityEngine.InputSystem.InputControl parent) 
 at UnityEngine.InputSystem.Layouts.InputDeviceBuilder.Setup (UnityEngine.InputSystem.Utilities.InternedString layout, UnityEngine.InputSystem.Utilities.InternedString variants, UnityEngine.InputSystem.Layouts.InputDeviceDescription deviceDescription)
```

Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidDevice.cs:18]: @ekcoh I have tried pretty much everything but I couldn't find a way to have no `ButtonControl`. This is the exception I am hitting ([code](https://github.com/Unity-Technologies/InputSystem/blob/platform/android/add-back-button-leaves-app-new-input/Packages/com.unity.inputsystem/InputSystem/Devices/InputDeviceBuilder.cs#L766)).

```
Could not create a device for 'AndroidDevice (Android)' (exception: System.InvalidOperationException: Control '/AndroidDevice' with layout 'AndroidDevice' has no size set and has no children to compute size from
 at UnityEngine.InputSystem.Layouts.InputDeviceBuilder.ComputeStateLayout (UnityEngine.InputSystem.InputControl control)
 at UnityEngine.InputSystem.Layouts.InputDeviceBuilder.InstantiateLayout (UnityEngine.InputSystem.Layouts.InputControlLayout layout, UnityEngine.InputSystem.Utilities.InternedString variants, UnityEngine.InputSystem.Utilities.InternedString name, UnityEngine.InputSystem.InputControl parent) 
 at UnityEngine.InputSystem.Layouts.InputDeviceBuilder.Setup (UnityEngine.InputSystem.Utilities.InternedString layout, UnityEngine.InputSystem.Utilities.InternedString variants, UnityEngine.InputSystem.Layouts.InputDeviceDescription deviceDescription)
```

There is in fact a test that check that a [DeviceWithNoControls](https://github.com/Unity-Technologies/InputSystem/blob/platform/android/add-back-button-leaves-app-new-input/Assets/Tests/InputSystem/CoreTests_Controls.cs#L1339) fails to be created: [Controls_CanTurnControlPathIntoHumanReadableText_EvenIfLayoutCannotBeFoundOrHasErrors](https://github.com/Unity-Technologies/InputSystem/blob/platform/android/add-back-button-leaves-app-new-input/Assets/Tests/InputSystem/CoreTests_Controls.cs#L1339)

File: Assets/Tests/InputSystem/CoreTests_Devices.cs
Comment (manugildev) [reply Assets/Tests/InputSystem/CoreTests_Devices.cs]: 84e41ff

File: Assets/Tests/InputSystem/CoreTests_Devices.cs
Comment (manugildev) [reply Assets/Tests/InputSystem/CoreTests_Devices.cs]: 84e41ff

File: Assets/Tests/InputSystem/Plugins/AndroidTests.cs
Comment (manugildev) [reply Assets/Tests/InputSystem/Plugins/AndroidTests.cs]: 84e41ff

File: Packages/com.unity.inputsystem/InputSystem/Devices/Commands/GetCustomCommand.cs
Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Devices/Commands/GetCustomCommand.cs:15]: The PR including the FourCC code in `unity/unity` is still on a draft, I am waiting for this InputSystem PR to be finalize so I can merge that one first. https://github.cds.internal.unity3d.com/unity/unity/pull/39251/files

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs
Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs:12]: `backButtonLeavesApp` is a legacy input setting. It's only living on Android land now. Even if we have the code for Windows platforms still living in our sourcecode, there are no supported device that has a physical back button anymore. They only had them in Windows Phones and early Windows RT tablets, none of which are supported anymore.

So it makes sense that moving innto the new Input System, we only added in the only device that still has a back button functionality, Android.

Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs:12]: @ekcoh `backButtonLeavesApp` is a legacy input setting. It's only living on Android land now, it is needed for `predictive back gesture` to work (which is something we have implemented for Android 14 as per Google's request). 

Even if we have the code for `Windows` platforms still living in unity/unity, there are no supported device that has a physical back button anymore. They only had them in Windows Phones and early Windows RT tablets, none of which are supported anymore. This has been confirmed by the windows-platform team.

So it makes sense that moving into the new Input System, we only added in the only device that still has a back button functionality, Android.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs
Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs:29]: 5a58763

Comment (manugildev) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/Android/InputSettingsAndroid.cs:29]: d4f551d

**Comment by manugildev** [comment]
(force push the only contains auto formatter changes)

**Comment by manugildev** [comment]
(force push the only contains auto formatter changes)

**Comment by manugildev** [comment]
@ekcoh I did some new changes based on the code review. Let me know if this is ready so I can proceed with landing the `unity/unity` PR first.

**Comment by ekcoh** [comment]
Sorry for providing this feedback at this stage, but I would suggest we do not land this PR for the following reasons:
- This API (backButtonLeavesApp), is a poor fit for a device representation. Only adds overhead and contains no streaming data and isn't tightly related to input.
- A corresponding API backButtonLeavesApp already exists in UnityEngine.Input and seems (in native implementation) to not be tied to InputManager at all so I see no reason to have two versions of the same API?

Please let me know or DM me if there is some value you see with this additional API in addition to what already exists in Unity.Input?

**Comment by seeyam-q** [comment]
@ekcoh  Hello I stumbled upon this thread as I was searching for the Input System's equivalent to `backButtonLeavesApp`. Because it is recommended to only have the new Input System enabled in the Unity project,  `backButtonLeavesApp` in UnityEngine.Input doesn't work anymore. 

It is very confusing to say "No corresponding API yet." in the Input System doc without providing a workaround to handle back button on Android, especially considering back button behavior is required by Google Play Store. 

If Unity is not supporting `backButtonLeavesApp` in the Input System for engineering reasons, can you at least provide a work around of handling back button in the Input System in the documentation?

**Comment by manugildev** [comment]
@seeyam-q thanks for bringing this up! Actually, last friday we merged a change to make `Input.backButtonLeavesApp` work on the new Input System on Android.

We are expecting these changes to land in `6000.0.14f1`, `2022.3.42f1` and `2021.3.43f1`.

→ [Issue Tracker Link `UUM-41638`](https://issuetracker.unity3d.com/product/unity/issues/guid/UUM-41638)

**Pull request closed without merging**
